warning: 'void sntp_setservername(u8_t, const char*)' is deprecated:

### The problem

Compiling .pioenvs/p-sz-ble-sensecapnightlight/src/esphome/components/socket/bsd_sockets_impl.cpp.o
src/esphome/components/sntp/sntp_component.cpp: In member function 'virtual void esphome::sntp::SNTPComponent::setup()':
src/esphome/components/sntp/sntp_component.cpp:34:21: warning: 'void sntp_setservername(u8_t, const char*)' is deprecated: use esp_sntp_setservername() instead [-Wdeprecated-declarations]
   34 |   sntp_setservername(0, strdup(this->server_1_.c_str()));
      |   ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from src/esphome/components/sntp/sntp_component.cpp:5:
/data/cache/platformio/packages/framework-espidf@3.50202.0/components/lwip/include/apps/esp_sntp.h:249:6: note: declared here
  249 | void sntp_setservername(u8_t idx, const char *server)
      |      ^~~~~~~~~~~~~~~~~~
src/esphome/components/sntp/sntp_component.cpp:36:23: warning: 'void sntp_setservername(u8_t, const char*)' is deprecated: use esp_sntp_setservername() instead [-Wdeprecated-declarations]
   36 |     sntp_setservername(1, strdup(this->server_2_.c_str()));
      |     ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/data/cache/platformio/packages/framework-espidf@3.50202.0/components/lwip/include/apps/esp_sntp.h:249:6: note: declared here
  249 | void sntp_setservername(u8_t idx, const char *server)
      |      ^~~~~~~~~~~~~~~~~~
src/esphome/components/sntp/sntp_component.cpp:39:23: warning: 'void sntp_setservername(u8_t, const char*)' is deprecated: use esp_sntp_setservername() instead [-Wdeprecated-declarations]
   39 |     sntp_setservername(2, strdup(this->server_3_.c_str()));
      |     ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/data/cache/platformio/packages/framework-espidf@3.50202.0/components/lwip/include/apps/esp_sntp.h:249:6: note: declared here
  249 | void sntp_setservername(u8_t idx, const char *server)
      |      ^~~~~~~~~~~~~~~~~~
src/esphome/components/sntp/sntp_component.cpp:45:12: warning: 'void sntp_init()' is deprecated: use esp_sntp_init() instead [-Wdeprecated-declarations]
   45 |   sntp_init();
      |   ~~~~~~~~~^~
/data/cache/platformio/packages/framework-espidf@3.50202.0/components/lwip/include/apps/esp_sntp.h:255:6: note: declared here
  255 | void sntp_init(void)
      |      ^~~~~~~~~

### Which version of ESPHome has the issue?

2024.11.0 and older

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

2024.11.2

### What platform are you using?

ESP32-IDF

### Board

esp32-s3-devkitc-1

### Component causing the issue

sntp / platformio/espressif32@6.7.0 framework-espidf @ 3.50202.0

### Example YAML snippet

```yaml
time:
  - platform: sntp
    id: sntp_time
    timezone: Europe/Zurich
    servers:
      - 192.168.166.1
    on_time_sync:
      - script.execute: time_update
    on_time:
      - minutes: '*'
        seconds: 0
        then:
          - script.execute: time_update

script:
  - id: time_update
    then:
      - lvgl.label.update:
          id: lvgl_clock_minute
          text:
            format: "%02d"
            args: [ 'id(sntp_time).now().minute' ]

      - lvgl.label.update:
          id: lvgl_clock_hour
          text:
            format: "%02d"
            args: [ 'id(sntp_time).now().hour' ]
      - lvgl.label.update:
          id: lvgl_clock_longdate
          text:
            format: "%s, %02d.%02d.%04d"
            args: 
              - '(new const char *[7]{"Sunntig", "Määntig", "Ziischtig", "Mittwuch", "Dunschtig", "Friitig", "Samschtig"})[id(sntp_time).now().day_of_week-1]'
              - 'id(sntp_time).now().day_of_month'
              - 'id(sntp_time).now().month'
              - 'id(sntp_time).now().year'
```


### Anything in the logs that might be useful for us?

_No response_

### Additional information

_No response_

--- a/esphome/components/sntp/sntp_component.cpp
+++ b/esphome/components/sntp/sntp_component.cpp
@@ -9,11 +9,6 @@
 #include "lwip/apps/sntp.h"
 #endif
 
-// Yes, the server names are leaked, but that's fine.
-#ifdef CLANG_TIDY
-#define strdup(x) (const_cast<char *>(x))
-#endif
-
 namespace esphome {
 namespace sntp {
 
@@ -26,30 +21,29 @@ void SNTPComponent::setup() {
     esp_sntp_stop();
   }
   esp_sntp_setoperatingmode(ESP_SNTP_OPMODE_POLL);
+  size_t i = 0;
+  for (auto &server : this->servers_) {
+    esp_sntp_setservername(i++, server.c_str());
+  }
+  esp_sntp_set_sync_interval(this->get_update_interval());
+  esp_sntp_init();
 #else
   sntp_stop();
   sntp_setoperatingmode(SNTP_OPMODE_POLL);
-#endif
 
-  sntp_setservername(0, strdup(this->server_1_.c_str()));
-  if (!this->server_2_.empty()) {
-    sntp_setservername(1, strdup(this->server_2_.c_str()));
+  size_t i = 0;
+  for (auto &server : this->servers_) {
+    sntp_setservername(i++, server.c_str());
   }
-  if (!this->server_3_.empty()) {
-    sntp_setservername(2, strdup(this->server_3_.c_str()));
-  }
-#ifdef USE_ESP_IDF
-  esp_sntp_set_sync_interval(this->get_update_interval());
-#endif
-
   sntp_init();
+#endif
 }
 void SNTPComponent::dump_config() {
   ESP_LOGCONFIG(TAG, "SNTP Time:");
-  ESP_LOGCONFIG(TAG, "  Server 1: '%s'", this->server_1_.c_str());
-  ESP_LOGCONFIG(TAG, "  Server 2: '%s'", this->server_2_.c_str());
-  ESP_LOGCONFIG(TAG, "  Server 3: '%s'", this->server_3_.c_str());
-  ESP_LOGCONFIG(TAG, "  Timezone: '%s'", this->timezone_.c_str());
+  size_t i = 0;
+  for (auto &server : this->servers_) {
+    ESP_LOGCONFIG(TAG, "  Server %zu: '%s'", i++, server.c_str());
+  }
 }
 void SNTPComponent::update() {
 #if !defined(USE_ESP_IDF)

--- a/esphome/components/sntp/sntp_component.h
+++ b/esphome/components/sntp/sntp_component.h
@@ -14,23 +14,20 @@ namespace sntp {
 /// \see https://www.gnu.org/software/libc/manual/html_node/TZ-Variable.html
 class SNTPComponent : public time::RealTimeClock {
  public:
+  SNTPComponent(const std::vector<std::string> &servers) : servers_(std::move(servers)) {}
+
+  // Note: set_servers() has been removed and replaced by a constructor - calling set_servers after setup would
+  // have had no effect anyway, and making the strings immutable avoids the need to strdup their contents.
+
   void setup() override;
   void dump_config() override;
-  /// Change the servers used by SNTP for timekeeping
-  void set_servers(const std::string &server_1, const std::string &server_2, const std::string &server_3) {
-    this->server_1_ = server_1;
-    this->server_2_ = server_2;
-    this->server_3_ = server_3;
-  }
   float get_setup_priority() const override { return setup_priority::BEFORE_CONNECTION; }
 
   void update() override;
   void loop() override;
 
  protected:
-  std::string server_1_;
-  std::string server_2_;
-  std::string server_3_;
+  std::vector<std::string> servers_;
   bool has_time_{false};
 };
 

--- a/esphome/components/sntp/sntp_component.h
+++ b/esphome/components/sntp/sntp_component.h
@@ -14,7 +14,7 @@ namespace sntp {
 /// \see https://www.gnu.org/software/libc/manual/html_node/TZ-Variable.html
 class SNTPComponent : public time::RealTimeClock {
  public:
-  SNTPComponent(const std::vector<std::string> &servers) : servers_(std::move(servers)) {}
+  SNTPComponent(const std::vector<std::string> &servers) : servers_(servers) {}
 
   // Note: set_servers() has been removed and replaced by a constructor - calling set_servers after setup would
   // have had no effect anyway, and making the strings immutable avoids the need to strdup their contents.

--- a/esphome/components/sntp/time.py
+++ b/esphome/components/sntp/time.py
@@ -1,16 +1,16 @@
+import esphome.codegen as cg
 from esphome.components import time as time_
 import esphome.config_validation as cv
-import esphome.codegen as cg
-from esphome.core import CORE
 from esphome.const import (
     CONF_ID,
     CONF_SERVERS,
+    PLATFORM_BK72XX,
     PLATFORM_ESP32,
     PLATFORM_ESP8266,
     PLATFORM_RP2040,
     PLATFORM_RTL87XX,
-    PLATFORM_BK72XX,
 )
+from esphome.core import CORE
 
 DEPENDENCIES = ["network"]
 sntp_ns = cg.esphome_ns.namespace("sntp")
@@ -40,11 +40,8 @@ CONFIG_SCHEMA = cv.All(
 
 
 async def to_code(config):
-    var = cg.new_Pvariable(config[CONF_ID])
-
     servers = config[CONF_SERVERS]
-    servers += [""] * (3 - len(servers))
-    cg.add(var.set_servers(*servers))
+    var = cg.new_Pvariable(config[CONF_ID], servers)
 
     await cg.register_component(var, config)
     await time_.register_time(var, config)

